feat(appmesh): add missing route match features

misterjoshua opened the pull request:
Adds route priority, header matching and matching by scheme and method.

Closes #11645

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 9417d0a52431bf699447d9369c23896eeb48a429
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=bf6fe14c-ad57-40f7-adb5-d7fa51378289%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 612de8dbbdd5564b62d4b223e189581f7e777c97
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=629357c1-89d1-4ecf-aa97-772815a05525%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: d930e921a80dbb9c2971ca0be4b8f0618909beec
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=52d5a2fb-a0c5-449e-abb1-54f11c09c5f0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 5766d5474426c0673a9673884b73cab1d4bcb3b5
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=81ebe345-e7dc-4cfd-9bd1-a553662b0dc7%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:
```diff
@@ -34,6 +34,179 @@ export interface HttpRouteMatch {
+  HTTPS = 'https',
+}
+
+/**
+ * An object that represents the HTTP header in the request to match.
+ */
+export interface HttpRouteHeader {
+  /**
+   * Specify `true` to match anything except the match criteria.
+   * @default false
+   */
+  readonly invert?: boolean;
+
+  /**
+   * The method to use to match the header.
+   * @default - no match method
+   */
+  readonly match?: IHeaderMatchMethod;
+
+  /**
+   * A name for the HTTP header in the client request that will be matched on.
+   * @example 'content-type'
+   */
+  readonly name: string;
+}
```
I gotta say, I'm not a huge fan of this API. Feels a little bit like too much ceremony for something simple. Here's how I would envision an API that I would like:

```ts
router.addRoute('route-http2', {
  routeSpec: appmesh.RouteSpec.http2({
    weightedTargets: [
      {
        virtualNode: node,
      },
    ],
    match: {
      prefixPath: '/my/path',
      headers: [
        appmesh.HeaderMatch.valueIs('Content-Type', 'application/json'),
        appmesh.HeaderMatch.valueStartsWith('Content-Type', 'application/json'),
        appmesh.HeaderMatch.valueEndsWith('Content-Type', 'application/json'),
        appmesh.HeaderMatch.valueMatchesRegex('Content-Type', 'application/.*'),
        appmesh.HeaderMatch.valueIsBetween('Content-Type', 1, 5),
        appmesh.HeaderMatch.valueIsNot('Content-Type', 'application/json'),
        appmesh.HeaderMatch.valueDoesNotStartWith('Content-Type', 'application/json'),
        appmesh.HeaderMatch.valueDoesNotEndWith('Content-Type', 'application/json'),
        appmesh.HeaderMatch.valueDoesNotMatchRegex('Content-Type', 'application/.*'),
        appmesh.HeaderMatch.valueIsNotBetween('Content-Type', 1, 5),
      ],
    },
  }),
});
```

What do you think of this idea @misterjoshua ?

misterjoshua replied:
@skinny85 This looks pretty good! In this PR, opted for the way @dfezzie suggested in #13350 only because I didn't have a strong enough opinion, so I'm easily convinced. Lol. I have enough to get started.

nit., I'm not sure about `valueIsBetween` because the range App Mesh wants is `[start, end)` (inclusive of start, exclusive of end) and "Between" has some linguistic closed interval baggage. With this in mind, how about `valueIsInRange`?

skinny85 replied:
Another point about the `headers` property: I assume the result is a logical AND of all of the conditions inside of the array? If so, it might be worth it to call it out explicitly in the documentation.

misterjoshua replied:
Ok. I've pushed up a docs change that mentions this.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:
```diff
@@ -34,6 +34,179 @@ export interface HttpRouteMatch {
+  readonly name: string;
+}
+
+/**
+ * Match method.
+ */
+export interface IHeaderMatchMethod {
+  /**
+   * Produce the match method.
+   */
+  renderMatchMethod(): CfnRoute.HeaderMatchMethodProperty;
+}
+
+/**
+ * Used to generate header matching methods.
+ */
+export abstract class HeaderMatchMethod {
```
Let's use the `bind()` pattern here instead.

1. Rename `IHeaderMatchMethod` to `HeaderMatchMethodConfig`, and make it have a `readonly matchProperty:  CfnRoute.HeaderMatchMethodProperty;` as its only property.
2. Create an abstract `bind(scope: Construct): HeaderMatchMethodConfig` method in `HeaderMatchMethod`.
3. Create a module-private implementation class that extends `HeaderMatchMethod` that is returned from each of the static factory methods in `HeaderMatchMethod` (whose return type should each be `HeaderMatchMethod`).
4. Call `bind()` in the implementation of  each `RouteSpec`.

The AppMesh module uses this pattern quite often, you can look at how it's implemented in other places for further guidance (for example, [for `ClientPolicy`](https://github.com/aws/aws-cdk/blob/ba99ad086e309fe5d11c75351e4b5dbb1f199478/packages/%40aws-cdk/aws-appmesh/lib/client-policy.ts#L8-L113)).

skinny85 reviewed (changes requested):
Generally looks great @misterjoshua !

Couple of minor comments/questions.

misterjoshua commented:
@skinny85 I've changed the header matching to your `HeaderMatch` bind pattern. I had to make a few judgement calls in type naming and `HeaderMatchConfig` properties. I went with `valueInRange` and `valueNotInRange` for the reason I mentioned above, but I have no strong objection to switching to `valueIsBetween` and `valueIsNotBetween` in case you have a strong opinion there. Let me know if there's anything else you need. Thanks

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 82e8096dae83f24f4b84a9d09693d3f71dd10095
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=62bcdf05-7336-4ef4-9ba6-97df59ae744e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 5258a65c4da9cd22c2906242bea20162a9abdfb6
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=d86608f6-3a13-42a8-89a2-596b3cfcf6ef%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 68c417137a4c70095df8a7404caff5c9d063442c
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=fbc18110-ae43-4afb-81f2-950dc20e35be%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

alexbrjo reviewed (approved):
LGTM!

misterjoshua commented on packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:
```diff
@@ -34,6 +34,223 @@ export interface HttpRouteMatch {
+  /**
+   * The value sent by the client must not include the specified characters.
+   */
+  static valueDoesNotMatchRegex(name: string, regex: string): HeaderMatch {
```
@alexbrjo Thanks for your feedback! Out of curiosity, what approach would you have taken?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 4107fd06788dac8d501c7dc609db0776988ba999
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=28693b25-d960-4c63-94b1-0a3d8b2cc9e6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:
```diff
@@ -34,6 +34,223 @@ export interface HttpRouteMatch {
+  readonly headers?: HeaderMatch[];
+
+  /**
+   * The client request method to match on. Specify only one.
```
What does "specify only one" means in this case?

I'd also mention explicitly this is the "HTTP request method", for clarity.

misterjoshua replied:
Thanks for catching this. "Specify only one" can be removed because we're using an enum instead of a string, so there's no need to be explicit in the docs that only one HTTP method can be matched on - it's a holdover from the [CFN docs](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-appmesh-route-httproutematch.html#cfn-appmesh-route-httproutematch-method).

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:
```diff
@@ -34,6 +34,223 @@ export interface HttpRouteMatch {
+   *
+   * @default - do not match on HTTP2 scheme
+   */
+  readonly scheme?: HttpRouteMatchScheme;
```
"Scheme" is a weird word to use here. How about "protocol" instead?

misterjoshua replied:
Yeah. It's just the IETF's choice of wording in the HTTP/2 and URI RFCs. It corresponds directly with the [`:scheme`](https://tools.ietf.org/html/rfc7540#section-8.1.2.3) pseudo-header which refers to the request's [scheme URI portion](https://tools.ietf.org/html/rfc3986#section-3.1). Notwithstanding, I'll push up a commit that changes `scheme` -> `protocol`. It's enough to change back if there's any desire.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:
```diff
@@ -34,6 +34,223 @@ export interface HttpRouteMatch {
+}
+
+/**
+ * Supported request matching methods.
```
This reads a little awkward. How about: "Supported values for matching routes based on the HTTP request method"?

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:
```diff
@@ -34,6 +34,223 @@ export interface HttpRouteMatch {
+  /**
+   * The Http header name
+   */
+  readonly name: string;
```
Let's call this `headerName` then.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:
```diff
@@ -34,6 +34,223 @@ export interface HttpRouteMatch {
+   * @param start Match on values starting at and including this value
+   * @param end Match on values up to but not including this value
+   */
+  static valuesIsNotInRange(name: string, start: number, end: number): HeaderMatch {
```
I feel like `name` should be `headerName` everywhere here.

That parameter is also missing the JSDocs for each of those methods.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:
```diff
@@ -34,6 +34,223 @@ export interface HttpRouteMatch {
+/**
+ * Supported :scheme options for HTTP2
+ */
+export enum HttpRouteMatchScheme {
```
I feel like `HttpRouteProtocol` is a better name here.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:
```diff
@@ -34,6 +34,223 @@ export interface HttpRouteMatch {
+ */
+export abstract class HeaderMatch {
+  /**
+   * The value sent by the client must match the specified value exactly.
```
This "The value sent by the client" is a weird phrasing I feel like.

How about "The value of the header with the given name in the request" instead?

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:
```diff
@@ -34,6 +34,223 @@ export interface HttpRouteMatch {
+  /**
+   * The value sent by the client must match the specified value exactly.
+   */
+  static valueIs(name: string, exact: string): HeaderMatch {
```
`exact` is a weird parameter name... How about just `headerValue`?

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:
```diff
@@ -209,30 +452,53 @@ class HttpRouteSpec extends RouteSpec {
       throw new Error(`Prefix Path must start with \'/\', got: ${prefixPath}`);
     }
+
+    let headers: CfnRoute.HttpRouteHeaderProperty[] | undefined;
```
Can we inline this variable instead below, where it's only usage is?

```ts
      match: {
        prefix: prefixPath,
        headers: this.match.headers?.map(header => {
          const config = header.bind(_scope);
          return {
            name: config.name,
            invert: config.invert,
            match: config.matchProperty,
          };
        }),
        method: this.match?.method,
        scheme: this.match?.scheme,
      },

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:
```diff
@@ -183,6 +421,11 @@ export abstract class RouteSpec {
 }
 
 class HttpRouteSpec extends RouteSpec {
+  /**
```
No need for the comments on these classes (they are module-private - no `export` keyword).

skinny85 reviewed (changes requested):
Looks great @misterjoshua ! Some minor documentation and naming changes that I feel like are needed before merging.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:
```diff
@@ -34,6 +34,223 @@ export interface HttpRouteMatch {
+   * The match property.
+   */
+  readonly matchProperty: CfnRoute.HeaderMatchMethodProperty;
+}
```
Instead of these 3 properties in this interface, can you instead have one `readonly httpRouteHeader: CfnRoute.HttpRouteHeaderProperty`?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: 88c24be5a6a0231f59ea8dd6f81d39ff27f602a4
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b6be7b38-2c65-4fbc-905f-b40447719186%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

misterjoshua commented:
@skinny85 Alright. I've pushed some commits that address your review items. Take a look and let me know if there's anything else you'd like. Thanks.

skinny85 commented on packages/@aws-cdk/aws-appmesh/README.md:
```diff
@@ -298,6 +298,30 @@ router.addRoute('route-http', {
+    match: {
+      prefixPath: '/path-to-app',
+      method: appmesh.HttpRouteMatchMethod.POST,
+      scheme: appmesh.HttpRouteMatchScheme.HTTPS,
```
I believe those are now out of the date, with the recent name changes.

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:
```diff
@@ -34,6 +34,255 @@ export interface HttpRouteMatch {
+/**
+ * Configuration for `HeaderMatch`
+ */
+export interface HeaderMatchConfig {
```
Can we rename this `HttpHeaderMatchConfig`?

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:
```diff
@@ -34,6 +34,255 @@ export interface HttpRouteMatch {
+/**
+ * Used to generate header matching methods.
+ */
+export abstract class HeaderMatch {
```
Can we rename this to `HttpHeaderMatch`?

skinny85 commented on packages/@aws-cdk/aws-appmesh/lib/route-spec.ts:
```diff
@@ -34,6 +34,255 @@ export interface HttpRouteMatch {
+   * The value of the header with the given name in the request must match the
+   * specified value exactly.
+   *
+   * @param headerName Test against the value of this HTTP request header
```
How about just

```suggestion
   * @param headerName the name of the HTTP header to match against
```

? 🙂

Same comment everywhere below.

skinny85 reviewed (commented):
This looks fantastic @misterjoshua !

Given that you've missed updating the ReadMe with the rename changes, I've snuck in a few last-minute naming suggestions - most importantly, I think we should add the `Http` prefix to the `HeaaderMatch` types, to make it clear those are only for HTTP headers.

Thanks for all your work on this!

misterjoshua commented:
> This looks fantastic @misterjoshua !
> 
> Given that you've missed updating the ReadMe with the rename changes, I've snuck in a few last-minute naming suggestions - most importantly, I think we should add the `Http` prefix to the `HeaderMatch` types, to make it clear those are only for HTTP headers.
> 
> Thanks for all your work on this!

Haha whoops. Nice catch. I'll take a closer at this tomorrow since I'll need to re-run the integ anyway, once #13353 is in.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: cb9012ea442b2659b8caa0304f7fba47f621228c
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=39a29bf3-c29b-415e-9165-0f3da2c1059f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

misterjoshua commented:
@skinny85 Hey. I renamed, fixed the param docs, and fixed the README. I also merged the latest from master and re-ran integ to make sure it all still deploys. Let me know if there's anything else!

skinny85 reviewed (approved):
Looks perfect @misterjoshua , thanks so much for the contribution!
